Basic Unit Tests

matwld opened the pull request:
Add basic unit tests, the coverage is about 50% of the code. Need to add more edge cases but would like review before getting deeper. Also show learnings on mocking in jest.

matwld committed:
chore(test): add skeleton unit tests

- add skeleton unit tests to be filled in later
- fix basic tests for deploy and init
- add --collectCoverage to test run command in package.json

matwld committed:
chore(tests): add model1.glb file

- add model1 glb file for reading/writing in unit tests

matwld committed:
fix(init): prefer s3 conversion to string, simplify workspaceId

- remove custom stream reading logic, replace with s3 native method for
  conversion to string
- simplify use of workspaceId by removing workspaceIdStr

matwld committed:
fix(tests): skeleton unit tests, implement init tests

- add skeleton test cases for each implemented command
- implement basic tests for init

matwld committed:
fix(tests): refactor unit tests

- refactor test constants to new file
- remove cast to 'any'

matwld committed:
fix(init test): remove casts to any

remove casts to any

matwld committed:
feat(nuke test): implement basic nuke unit tests

implement basic nuke unit tests

matwld committed:
chore(audit test): disable audit test

disable audit test until command is implemented

matwld committed:
chore(jest): update jest version

update jest version

matwld committed:
chore(deploy): throw error for workspace not found

throw error for workspace does not exist, will be changed later

matwld committed:
feat(deploy tests): implement basic deploy unit tests

implement basic deploy unit tests

matwld committed:
chore(tests): rename variables and clean up unit tests

clean up basic unit tests

matwld committed:
fix(commands): standardize verify workspace does not exist

- standardize use of verifyWorkspaceExists
- make verify method throw ResourceNotFoundException
- standardize workspaceId variable in commands

matwld committed:
fix(unit tests): update tests

- update tests to use component_types
- update expectation of workspaceNotFound case
- add mock for verify tmdk.json exists

matwld committed:
fix(eslint): resolve eslint errors

- resolve eslint errors
- remove some unncessary imports

there are still some warnings about non-undefined assertions, will need
to revisit

diehbria commented on src/libs/tmdk/tests/deploy.test.ts:
```diff
@@ -1,26 +1,265 @@
+    mockExit.mockClear();
+  });
 
   test("deploy_givenNoTMDKProject_expectError", async () => {
```
prefer test names such as 'test('throws error when given directory that does not exist')' i.e. human readable test descriptions in the for mat of "does X when Y"

matwld replied:
Got it, will update

diehbria commented on src/libs/tmdk/tests/deploy.test.ts:
```diff
@@ -1,26 +1,265 @@
+const s3Mock = mockClient(S3Client);
+const fakeTmdkDir = "/tmp/deploy-unit-tests";
 
 describe("testing deploy", () => {
```
all tests are always testing, so no need to say that the test is testing

matwld replied:
Ok, for test suites how do we prefer to name them?

diehbria replied:
Typically use `describe` block to logically group tests, for example, in this test suite: https://github.com/awslabs/iot-app-kit/blob/main/packages/core/src/common/dataFilters.spec.ts

each function being tested in a file that exports several, has a describe block.

For your use case, i would probably do something such as:


describe('deploys succesfully')
and

describe('fails to deploy')

as a way to categorize your test suites.

Since your deploy tests are contained in a deploy test file, it's already implied (and clear in the jest test runner results) that they are all related to deployment, so a single file level wide describe('deploys') adds no information though its a common pattern

diehbria commented on src/libs/tmdk/tests/deploy.test.ts:
```diff
@@ -1,26 +1,265 @@
+  });
+
+  test("deploy_given1ComponentType_expectSuccess", async () => {
+    fs.existsSync = jest.fn(() => {
```
stylistic nit: () => true

matwld replied:
Got it, did not know that would work. Thanks!

diehbria commented on src/libs/tmdk/tests/deploy.test.ts:
```diff
@@ -1,26 +1,265 @@
+    );
+    expect(twinmakerMock.commandCalls(CreateEntityCommand).length).toBe(0);
+    expect(twinmakerMock.commandCalls(CreateSceneCommand).length).toBe(1);
+    expect(
```
there's a much cleaner way to assert on function called, using matchers such as `toHaveBeenCAlled() or `toHaveBeenLastCalledWith()`: https://jestjs.io/docs/mock-functions#custom-matchers

matwld replied:
I don't think this works for the aws mocking as I used the aws-sdk-mock package, not jest. I could not get the jest mock for s3 to work and for twinmaker it was unwieldy. Would be happy to discuss offline

diehbria commented on src/libs/tmdk/tests/deploy.test.ts:
```diff
@@ -1,26 +1,265 @@
+    expect(twinmakerMock.commandCalls(CreateSceneCommand).length).toBe(0);
+  });
+
+  test("deploy_given1SceneAnd1Model_expectSuccess", async () => {
```
better test name: `it('deploys successfully when there is one scene and one model')`

matwld replied:
Got it, will update

diehbria commented on src/libs/tmdk/tests/init.test.ts:
```diff
@@ -1,22 +1,59 @@
+/* eslint @typescript-eslint/no-var-requires: "off" */
```
if you have a very strong reason to disable a rule, better to disable it for a single line as opposed to the entire file

matwld replied:
I removed the code that required the rule to be disabled

matwld committed:
fix(unit tests): resolve pr feedback

resolve pr feedback by:
- changing test names to readable format
- removing description blocks from test files
- stylistic () => true
- remove mockExit and remove eslint rule disabling

matwld committed:
chore(configuration): update configuration to get closer to appkit

- change ts and jest configuration to align closer to appkit
- resolve isolatedModules issue in component-type and entity

johnnyw-aws commented on src/libs/tmdk/jest.config.js:
```diff
@@ -6,4 +6,12 @@ module.exports = {
   moduleFileExtensions: ["ts", "tsx", "js", "jsx", "json", "node"],
+  coverageThreshold: {
+    global: {
+      statements: 80,
+      branches: 80,
+      functions: 80,
+      lines: 80,
```
Is this too high for an alpha product? (maybe start with 50/60% if at all?)

Is there a way for us to emphasize integration test coverage instead of unit test coverage?

diehbria replied:
we want to have coverage threshold for every package - at a minimum, i want my team to be able to upgrade dependencies for security and package health reasons with 100% confidence nothing breaks. What code coverage is necessary to enable an away team to perform routine package updates without concern that something may regress?

matwld replied:
This is from appkit's coverage targets. The tests do not fail if we do not meet these targets, it just gives a warning. We can discuss offline about realistic coverage targets.

johnnyw-aws replied:
Thanks Brian, given that this tool is in an "early adoption / alpha phase" I'd like our acceptance testing to essentially be driven by correctness as defined through our end-to-end functional tests. Let's discuss offline on the coverage targets

johnnyw-aws commented on src/libs/tmdk/tests/deploy.test.ts:
```diff
@@ -3,24 +3,246 @@
+  expect(twinmakerMock.commandCalls(CreateSceneCommand).length).toBe(0);
+});
+
+test("deploys successfully when given a tmdk project with one component type", async () => {
```
Does this duplicate coverage from the integ test?

If so, would remove + de-prioritize happy-case coverage in unit tests and focus on negative cases. Mainly because I'd say the overall project is still in an alpha state so fine-grained happy cases may just cause us to have to update more places if things in the codebase change

matwld replied:
This does duplicate case from the functional test but I feel these tests are better for narrowing down specific cases where code fails. Functional test is more to verify that tmdk interacts with aws correctly. Also if code changes change the underlying use of aws api calls I think it is reasonable to expect developer to update the test cases as well.

Say for instance we change how deploy verifies whether a component type was created by using list component types after trying to create all of them instead of verifying each one in turn. While it will be additional work to change the test case to match this new behavior, without this fine grained test there is greater chance developer misses some edge case with their refactor.
